### DISCUSSION TITLE
Watching form for validity

### DISCUSSION THREAD

**Discussion by Bond-Addict** [epicmaxco/vuestic-ui#3335] [Q&A]
Does anyone know if its possible to watch the va-form for validity? I know you can manually call `.validate()` after creating a ref for the form, but I was wanting to have behavior like I could do in angular. If the form isnt valid I would like to either hide the submit button or disable it. So far I don't see a way in the docs. In angular you could just check for `formGroup.invalid` or `formGroup.valid` which would update automatically when all validation conditions were satisfied. 

TIA!

**Comment by m0ksem** [comment]
Hi, @Bond-Addict. You can try to use `useForm` composable (available since v1.6.4).

```html
<va-form ref="myForm" />
```

```ts
const { isValid } = useForm('myForm')

watch(isValid, console.log)
```

**Reply by Bond-Addict** [reply]
So using your answer I get: 

``Expected 2 arguments, but got 1.ts(2554)
useForm.d.ts(36, 76): An argument for 'props' was not provided.``

So I understand I also need to provide ExtractPropTypes but I'm not sure how to provide it.

Can you provide the docs on composables? I can't seem to find it.

Thanks

**Reply by m0ksem** [reply]
Double check if you have correct import name.
Must be `import { useForm } from 'vuestic-ui'`

https://vuestic.dev/en/ui-elements/form#default-usage

**Reply by Bond-Addict** [reply]
> Double check if you have correct import name. Must be `import { useForm } from 'vuestic-ui'`
> 
> https://vuestic.dev/en/ui-elements/form#default-usage

I just copied 
```
  import { useForm } from 'vuestic-ui'

  const { isValid, validate, reset, resetValidation } = useForm('formRef')
```
from https://vuestic.dev/en/ui-elements/form#default-usage 
and got
Module '"vuestic-ui"' has no exported member 'useForm'.

**Reply by Bond-Addict** [reply]
> Double check if you have correct import name. Must be `import { useForm } from 'vuestic-ui'`
> 
> https://vuestic.dev/en/ui-elements/form#default-usage


Any update on things I can try?

This is my package.json

```
	"dependencies": {
		"@mdi/font": "7.0.96",
		"@popperjs/core": "^2.11.6",
		"@vuestic/tailwind": "^0.1.3",
		"axios": "^1.3.4",
		"bootstrap": "^5.2.3",
		"core-js": "^3.8.3",
		"pinia": "^2.0.23",
		"popper.js": "^1.16.1",
		"roboto-fontface": "*",
		"vue": "^3.2.13",
		"vue-cli-plugin-vuestic-ui": "^1.0.11",
		"vue-router": "^4.0.0",
		"vuestic-ui": "^1.6.3",
		"webfontloader": "^1.0.0"
	},
	"devDependencies": {
		"@types/bootstrap": "^5.2.6",
		"@types/node": "^18.15.5",
		"@types/webfontloader": "^1.6.35",
		"@vitejs/plugin-vue": "^3.0.3",
		"sass": "^1.59.3",
		"typescript": "^4.9.5",
		"vite": "^3.2.5",
		"vue-tsc": "^1.0.9"
	}
```

**Reply by m0ksem** [reply]
Hi @Bond-Addict, try to update your vuestic-ui package to ^1.6.4 from ^1.6.3 :)

**Reply by Bond-Addict** [reply]
Hey @m0ksem, same issue. Do I possibly need to change something in my tsconfig?

This is what I currently have:

```
{
  "compilerOptions": {
    "baseUrl": ".",
    "target": "esnext",
    "useDefineForClassFields": true,
    "allowSyntheticDefaultImports": true,
    "composite": true,
    "module": "esnext",
    "moduleResolution": "node",
    "strict": true,
    "jsx": "preserve",
    "sourceMap": false,
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "lib": ["esnext", "dom"],
    "types": [
      "node",
    ],
    "paths": {
      "@/*": ["src/*"]
    },
    "isolatedModules": false,
    "noImplicitAny": false
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.d.ts",
    "src/**/*.tsx",
    "src/**/*.vue",
    "vite.config.ts"
  ],
}

```

**Reply by m0ksem** [reply]
> Hey @m0ksem, same issue. Do I possibly need to change something in my tsconfig?

I don't think problem in TS config. Try to remove node_modules and install them again if you have some caching issues.

Here is an example of a project that uses this feature:
https://codesandbox.io/p/sandbox/relaxed-roentgen-g923yx?file=%2Fsrc%2FApp.vue&selection=%5B%7B%22endColumn%22%3A25%2C%22endLineNumber%22%3A85%2C%22startColumn%22%3A25%2C%22startLineNumber%22%3A85%7D%5D

**Reply by Bond-Addict** [reply]
@m0ksem after clearing my node modules I don't get build errors. I do however still have issues when trying to use the ref. I downloaded your example and it does work, but on my project it doesn't seem to  pickup changes from the form. I tried to change a few things on your example such as changing your reactive variable to a ref. I also removed and additional form ref that I no longer needed since ``useForm`` has ``validate()``. My method is a bit more complex so maybe its something due to nesting?

I know at the least its still partially due to ``useForm`` not working. When logging ``isValid`` and ``validate()`` in my submit, ``isValid`` is always false and ``validate()`` is undefined.

Table has a modal with a slot which is where the form goes. I'm wanting to disable or hide the Ok button if invalid. That will be my next part to tackle after the ``isValid`` part is worked out.

Script:
```
const save = () => {
  console.log(`Before: ${isValid.value}`);
  console.log(`Called validate: ${validate()}`);
  if (validate()) {
    console.log(`After: ${isValid.value}`);
```
Template:
```
<template>
  <Table :table-params="tableParams" @edit="edit" @save="save" ref="table">
    <va-form class="grid grid-cols-2 md:grid-cols-3 gap-2" ref="form">
```

I also just ran another test. I see that ``useForm`` should accept either a string or a ``ref<typeof VaForm>``. I've tried to either pass a ``Ref<any>`` or even typing the ``ref`` by either ``ref<VaForm>()``, ``ref({} as VaForm)`` or ``ref({type: typeof VaForm})``.

**Reply by m0ksem** [reply]
@Bond-Addict, can I ask you to try to do the same you're doing in your app in codesandbox? I'm not able to reproduce your issue...

**Reply by Bond-Addict** [reply]
@m0ksem https://codesandbox.io/p/sandbox/nice-darwin-iqjx2m?file=%2Ftsconfig.json&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A31%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A31%7D%5D

**Reply by Bond-Addict** [reply]
@m0ksem did that sandbox work?

**Reply by Bond-Addict** [reply]
Hey @m0ksem, just wanted to check in and make sure that sandbox worked for you.

Thanks!

**Reply by m0ksem** [reply]
> Hey @m0ksem, just wanted to check in and make sure that sandbox worked for you.
> 
> Thanks!

Hi. Sorry, I had examination session and wasn't able to take a look! Will check this week. Thanks!

**Reply by Bond-Addict** [reply]
That's alright.

Thanks for the update! :)

**Reply by Bond-Addict** [reply]
Hey @m0ksem, just wanted to check if you'd gotten a chance to check this out yet. Thanks!

**Reply by m0ksem** [reply]
Hi, @Bond-Addict. Try to change code to:

```ts
const form = ref()

const { isValid, validate, reset, resetValidation } = useForm(form);
```

I think using string doesn't work in script setup... Will do a new issue.

**Reply by Bond-Addict** [reply]
> Hi, @Bond-Addict. Try to change code to:
> 
> ```ts
> const form = ref()
> 
> const { isValid, validate, reset, resetValidation } = useForm(form);
> ```
> 
> I think using string doesn't work in script setup... Will do a new issue.

So I have implemented this change. Although ``isValid`` seems to now show correct validity when calling ``validate()``, but ``isValid`` doesn't seem to be reactive, or consistently reactive. I've tried passing the computed ref a prop to my child and watching the prop value or the effect of the value. I've also tried watching ``isValid`` in the parent and then just calling a method on the child and passing the value. I would expect (and I am newer to Vue so I could be wrong) that one of these methods would allow me to hide the Ok button for the modal based on the validity. When being called, this method does what I need it too, but I cant trigger it based on the form validity. 

```
const setModalValidity = (isValid: boolean) => {
  let okayButton = document.querySelector('.va-modal__footer .va-button--normal:nth-of-type(2)') as HTMLButtonElement;
  if (okayButton) {
    if (isValid)
      okayButton.style.visibility = 'visible';
    else
      okayButton.style.visibility = 'hidden';
  }
}
```

**Reply by Bond-Addict** [reply]
@m0ksem 

Just wanted to check back in.

Thanks!

**Reply by m0ksem** [reply]
> @m0ksem
> 
> Just wanted to check back in.
> 
> Thanks!

You need to write something like:
```ts
watch(isValid, (newVal) => {
  okBtn.style.visibility = newVal ? 'visible' : 'hidden'
})
```

But better will be to use template syntax:
```vue
<va-modal>
  <template #cotnent="{ ok }">
     <va-card>
        <va-card-content> Form here </va-card-content>
        
        <va-card-actions>
           <va-button v-show="isValid">Ok</va-button>
        </va-card-actions>
     </va-card>
  </template>
</va-modal>
```

**Reply by Bond-Addict** [reply]
This is basically what I was doing and it doesn't work the way I would expect based on other frameworks (such as angular(FormGroup.valid or FormGroup.invalid)).

I want to hide the ok button (as soon as the modal is open and only if the form data is invalid) rather than waiting for the user to update a field causing the validation event. 

The problem is that ``isValid`` doesn't emit a value once the form is shown. Only once the form has been updated and rule has been violated. I've tried both watch and watchEffect and neither has produced the intended outcome. 

If I open a record with prepopulated data, I wouldnt want to hide the button because the form is valid. This is the real crux of the issue.

I've updated my sandbox to reflect what I'm referring to.

https://codesandbox.io/p/sandbox/nice-darwin-iqjx2m

This isn't an absolute necessity if there is no way around it, but this would greatly improve the user experience for my clients and mirror the current design for forms within their existing application ecosystem.

**Reply by Bond-Addict** [reply]
Hey @m0ksem, did that updated sandbox help provide some clarity on the issue?

**Reply by m0ksem** [reply]
> Hey @m0ksem, did that updated sandbox help provide some clarity on the issue?

Hi! Sorry for long response. You can use `immediate` prop and form will be validated right after rendering.

```vue
<va-form immediate />
```

Here is a rich example: https://ui.vuestic.dev/en/ui-elements/form#named-fields

**Reply by Bond-Addict** [reply]
This is definitely closer to what I am trying to accomplish.
Implementing this does validate immediately, but if its a new record, I'm telling my user they messed up before they even get a chance to interact with the form. 
 Is it possible to suppress validation messages when the form is pristine?
 
 Is there a concept of pristine, dirty, touched ect within form?

**Reply by Bond-Addict** [reply]
Hey @m0ksem, just wanted to check if you have an update. Thanks

**Reply by m0ksem** [reply]
> This is definitely closer to what I am trying to accomplish. Implementing this does validate immediately, but if its a new record, I'm telling my user they messed up before they even get a chance to interact with the form. Is it possible to suppress validation messages when the form is pristine?
> 
> Is there a concept of pristine, dirty, touched ect within form?

It is not possible now. Here is the issue: https://github.com/epicmaxco/vuestic-ui/issues/3449

**Reply by Bond-Addict** [reply]
@m0ksem okay gotcha. Thanks for opening the issue to be worked on. 
Apologies for all the issues and discussions on the library. I'm building a new app for work using this library as the only UI and I genuinely do  enjoy everything yall have put together as it is far more complete and stable than other Vue UI alternatives.

**Reply by m0ksem** [reply]
> @m0ksem okay gotcha. Thanks for opening the issue to be worked on. Apologies for all the issues and discussions on the library. I'm building a new app for work using this library as the only UI and I genuinely do enjoy everything yall have put together as it is far more complete and stable than other Vue UI alternatives.

Nice to hear that! 
In case you want to see your feature requests faster in UI lib, you can help us to make a requirements right in this issue (I mean write down exactly what feature must do and how it must work). And obviously you can just take this issue and make whatever you need, so I can publish a new version for you as soon as you make everything you need.

**Reply by Mini-ghost** [reply]
@Bond-Addict 

Hi!

I'm not sure if you'd like to try a third-party form validation library, such as Vorms.

I attempted to write a Vuestic UI example that incorporates form validation with Vorms. Vorms provides a `dirty` feature to check if the current values are equal to the initial values, which might meet your requirements.

[Vorms Example - With Vuestic UI](https://stackblitz.com/edit/vorms-vuestic-ui?file=src%2FApp.vue)

**Reply by Bond-Addict** [reply]
@Mini-ghost thanks for the suggestion! I did check it out and it does look  good! I appreciate the plug as I see you're one of the contributors lol I might end up using this. Ultimately it would be nice for more advanced validation to implemented into Vuestic its self. There is at least a team discussion started #3502. Maybe this will be another option that they could implement. 
Is there any advantage that you see using Vorms over Vuelidate? This was another suggestion that was given.

Regardless, nice work!

**Reply by Mini-ghost** [reply]
@Bond-Addict 

Thank you for your approval.

If developing with Vuestic UI and Vorms, I highly recommend using [Smart Form Components](https://vorms.mini-ghost.dev/advanced/smart-form-component.html). By encapsulating the form field components of Vuestic UI, it enhances usability and convenience. This approach simplifies the form handling process, improves development efficiency, and helps maintain clean and manageable code.

Since I'm not sufficiently familiar with the intricacies of using Vuelidate, I cannot provide an objective opinion. However, the most significant difference is that Vorms does not come with built-in rules, but it does offer integration with third-party schema validation libraries like yup and zod.

Vuelidate is indeed an excellent validation library, with its API syntax being completely distinct from Vorms. Ultimately, the choice between the two comes down to personal preference.
